it/src/main/java/com/google/cloud/teleport/it/bigtable/DefaultBigtableResourceManager.java

```diff
@@ -0,0 +1,429 @@
+ * a new testId will be formed for naming: {first 21 chars of long testId} + “-” + {8 char hash of
+ * testId}.
+ *
+ * <p>The class is thread-safe.
+ */
+public class DefaultBigtableResourceManager implements BigtableResourceManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultBigtableResourceManager.class);
+
+  private static final int MIN_INSTANCE_ID_LENGTH = 6;
+  private static final int MAX_INSTANCE_ID_LENGTH = 30;
+  private static final Pattern ILLEGAL_INSTANCE_CHARS = Pattern.compile("[\\W_]");
+
+  private static final String DEFAULT_CLUSTER_ZONE = "us-central1-a";
+  private static final int DEFAULT_CLUSTER_NUM_NODES = 1;
+  private static final StorageType DEFAULT_CLUSTER_STORAGE_TYPE = StorageType.SSD;
+
+  private final String projectId;
+  private final String instanceId;
+  private final String testId;
+  private final BigtableResourceManagerClientHandler bigtableResourceManagerClientHandler;
+
+  private boolean hasInstance = false;
+
+  @VisibleForTesting
+  DefaultBigtableResourceManager(
+      String testId, String projectId, CredentialsProvider credentialsProvider) throws IOException {
+    this(
+        DefaultBigtableResourceManager.builder(testId, projectId)
+            .setCredentialsProvider(credentialsProvider));
+  }
+
+  @VisibleForTesting
+  DefaultBigtableResourceManager(
+      String testId,
+      String projectId,
+      BigtableResourceManagerClientHandler bigtableResourceManagerClientHandler,
+      CredentialsProvider credentialsProvider)
+      throws IOException {
+    this(
+        bigtableResourceManagerClientHandler,
+        DefaultBigtableResourceManager.builder(testId, projectId)
+            .setCredentialsProvider(credentialsProvider));
+  }
+
+  private DefaultBigtableResourceManager(DefaultBigtableResourceManager.Builder builder)
+      throws IOException {
+    this(
+        new BigtableResourceManagerClientHandler(
+            builder.bigtableInstanceAdminSettings.build(),
+            builder.bigtableTableAdminSettings.build(),
+            builder.bigtableDataSettings.build()),
+        builder.testId,
+        builder.projectId,
+        builder.instanceId);
+  }
+
+  private DefaultBigtableResourceManager(
+      BigtableResourceManagerClientHandler bigtableResourceManagerClientHandler,
+      DefaultBigtableResourceManager.Builder builder) {
+    this(
+        bigtableResourceManagerClientHandler,
+        builder.testId,
+        builder.projectId,
+        builder.instanceId);
+  }
+
+  private DefaultBigtableResourceManager(
+      BigtableResourceManagerClientHandler bigtableResourceManagerClientHandler,
+      String testId,
+      String projectId,
+      String instanceId) {
+    this.projectId = projectId;
+    this.instanceId = instanceId;
+    this.testId = testId;
+    this.bigtableResourceManagerClientHandler = bigtableResourceManagerClientHandler;
+
+    // Check that the project and instance ID's conform to GCS standards
+    if (!isValidId(
```

Comment:
Makes sense to make it a separate method `ResourceManagerUtils.checkValidProjectId(String projectId) throws IllegalArgumentException` and move all constants like `MIN_PROJECT_ID_LENGTH` there. And then it will be a 1-liner here (and in other resource managers).

Reply:
I agree. I refactored to support this.